### PR TITLE
Feature/handle extracted compiled files

### PR DESCRIPTION
**Pull request by liamqma**

### CODE REVIEW

**Comment by liamqma** [comment]
- [ ] Test parcel-transformer in jira-frontend
- [ ] Test webpack-loader in jira-frontend

File: packages/babel-plugin-strip-runtime/src/index.ts [hunk 1]
```diff
@@ -47,6 +51,44 @@ export default declare<PluginPass>((api) => {
...
+              this.cwd,
+              this.opts.extractStylesToDirectory.dest,
+              dirname(relativePath),
+              cssFilename
+            );
+            mkdirSync(dirname(cssFilePath), { recursive: true });
+            writeFileSync(cssFilePath, sort(this.styleRules.sort().join('\n')));
+
+            // Add css import to file
+            path.unshiftContainer('body', template.ast(`import "./${cssFilename}";`));
```
Comment (dddlr) [inline packages/babel-plugin-strip-runtime/src/index.ts:90]: Can we guarantee that cssFilename doesn't contain any literal double quotes? (or any other character that is valid in a filename, but invalid in an import statement)

Comment (JakeLane) [reply packages/babel-plugin-strip-runtime/src/index.ts:90]: I think we'd be safe to just escape the double quotes `.replaceAll('"', '\\"')`

File: packages/babel-plugin-strip-runtime/src/types.ts [hunk 2]
```diff
@@ -10,6 +10,11 @@ export interface PluginOptions {
    * When set will prevent additional require (one import per rule) in the bundle.
    */
   compiledRequireExclude?: boolean;
+
+  /**
+   * When set, extract styles to an external CSS file
+   */
+  extractStylesToDirectory?: { source: string; dest: string };
```
[the remark this answers is not in the document]
Comment (JakeLane) [reply packages/babel-plugin-strip-runtime/src/types.ts:17]: Just TypeScript, which is useful when calling the plugin programatically

**Review by JakeLane — CHANGES_REQUESTED** [review]

File: packages/webpack-loader/src/__tests__/extract-plugin.test.ts [hunk 3]
```diff
@@ -182,4 +182,22 @@ describe('CompiledExtractPlugin', () => {
       "
     `);
   }, 10000);
+
+  it('should hanlde extracted styles', async () => {
```
[the remark this answers is not in the document]
Comment (liamqma) [reply packages/webpack-loader/src/__tests__/extract-plugin.test.ts:186]: ah, good catch

File: fixtures/babel-component-extracted/src/index.jsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { styled, css } from '@compiled/react';
+
+const Button = styled.button`
```
[the remark this answers is not in the document]
Comment (liamqma) [reply fixtures/babel-component-extracted/src/index.jsx:3]: yup, you are right.

File: packages/babel-plugin-strip-runtime/src/index.ts (hunk 1 above)
Comment (dddlr) [reply packages/babel-plugin-strip-runtime/src/index.ts:90]: It's just double quotes and backslashes that we need to be worried about, right?

File: packages/babel-plugin-strip-runtime/src/index.ts [hunk 5]
```diff
@@ -47,6 +51,44 @@ export default declare<PluginPass>((api) => {
               // If we used ESM it would blow up with CJS source, unfortunately.
             });
           }
+
+          if (this.opts.extractStylesToDirectory && this.styleRules.length > 0) {
+            // Build filename of the css file
+            const currentFilename = basename(filename);
+            const cssFilename = `${currentFilename.substring(
```
[the remark this answers is not in the document]
Comment (liamqma) [reply packages/babel-plugin-strip-runtime/src/index.ts]: Good catch. Has updated it.

File: packages/babel-plugin-strip-runtime/src/index.ts (hunk 1 above)
Comment (liamqma) [reply packages/babel-plugin-strip-runtime/src/index.ts:89]: Has escaped the invalid.

File: packages/babel-plugin-strip-runtime/src/index.ts [hunk 6]
```diff
@@ -53,17 +53,16 @@ export default declare<PluginPass>((api) => {
           }
 
           if (this.opts.extractStylesToDirectory && this.styleRules.length > 0) {
-            // Build filename of the css file
-            const currentFilename = basename(filename);
-            const cssFilename = `${currentFilename.substring(
-              0,
-              currentFilename.lastIndexOf('.')
-            )}.compiled.css`;
```
Comment (dddlr) [inline packages/babel-plugin-strip-runtime/src/index.ts]: Ah is there protection against accidental overwriting of an existing `.compiled.css` file?

For example, if we have `app+v1.ts` and `app=v1.ts`, would the generated `.compiled.css` file for the second file overwrite the generated `.compiled.css` file for the first file?

File: packages/babel-plugin-strip-runtime/src/index.ts [hunk 7]
```diff
@@ -53,17 +53,16 @@ export default declare<PluginPass>((api) => {
...
+            // Build and sanitize filename of the css file
+            const cssFilename = `${parse(filename).name}.compiled.css`.replace(
```
Comment (dddlr) [inline packages/babel-plugin-strip-runtime/src/index.ts:57]: Ah is there protection against accidental overwriting of an existing .compiled.css file?

For example, if we have app+v1.ts and app=v1.ts, would the generated .compiled.css file for the second file overwrite the generated .compiled.css file for the first file?

Comment (liamqma) [reply packages/babel-plugin-strip-runtime/src/index.ts:57]: I haven't thought about this. Good point. Let me write some code to protect this.

Comment (dddlr) [reply packages/babel-plugin-strip-runtime/src/index.ts:57]: No prob

Up to you how to handle it, but even something simple like erroring out should be ok i think

File: packages/babel-plugin-strip-runtime/src/index.ts [hunk 8]
```diff
@@ -86,7 +87,7 @@ export default declare<PluginPass>((api) => {
             writeFileSync(cssFilePath, sort(this.styleRules.sort().join('\n')));
 
             // Add css import to file
-            path.unshiftContainer('body', template.ast(`import "./${cssFilename}";`));
+            path.unshiftContainer('body', template.ast(`import "./${addslashes(cssFilename)}";`));
```
[the remark this answers is not in the document]
Comment (liamqma) [reply packages/babel-plugin-strip-runtime/src/index.ts:90]: oh yea, that's a good idea.

**Review by JakeLane — APPROVED** [review]

**Review by dddlr — APPROVED** [review]

**Pull request merged**